### PR TITLE
Convert Nitrox to using BepInEx 5.4.11

### PR DESCRIPTION
**Pull request by MrPurple6411**
Changes in this PR.

Changed QModDetector to detect the QMM dll instead of just the QMods folder
Fixed 2 transpilers that failed when converted to HarmonyX
Converted EntryPatch to:
 - Install BepInEx if not found.
 - Move QMM to a backup folder if found when launching multiplayer
 - Move QMM from backup on exiting launcher or when launching SinglePlayer
 - Copy the newly modified bootloader when playing MP or delete it when Exiting Launcher or playing SP

~~Alpha 1.4.0.0 Bepinex 5.4.1.1.zip~~
116cb5d - [Alpha 1.4.0.0 Bepinex 5.4.1.1.zip](https://github.com/SubnauticaNitrox/Nitrox/files/6587003/Alpha.1.4.0.0.Bepinex.5.4.1.1.zip)

### CODE REVIEW

File: NitroxLauncher/BepInEx/BepInEx/config/BepInEx.cfg [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+## The local filename of the assembly to target.
+# Setting type: String
+# Default value: UnityEngine.CoreModule.dll
+Assembly = Assembly-CSharp.dll
+
+## The name of the type in the entrypoint assembly to search for the entrypoint method.
+# Setting type: String
+# Default value: Application
+Type = SystemsSpawner
```
Comment (Jannify) [inline NitroxLauncher/BepInEx/BepInEx/config/BepInEx.cfg:11]: Why the change? What's the difference between `SystemsSpawner` and `GameInput`?

**Review by Jannify — CHANGES_REQUESTED** [review]
`NitroxTest` still depends on `Lib.Harmony` and not Harmony from BepInEx.
I would also move the BepInEx folder in the root because three projects depend on it.

I also can't get the code error free because `BaseUnityPlugin` is missing the reference to Unity, but I don't know if this is just for me because vs or whatever.

File: Nitrox.Bootloader/Nitrox.Bootloader.csproj [hunk 2]
```diff
@@ -1,34 +1,48 @@
...
+    </Reference>
+    <Reference Include="UnityEngine">
+      <HintPath>..\..\..\..\Program Files (x86)\Steam\steamapps\common\Subnautica\Subnautica_Data\Managed\UnityEngine.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+    <Reference Include="UnityEngine.CoreModule">
+      <HintPath>..\..\..\..\Program Files (x86)\Steam\steamapps\common\Subnautica\Subnautica_Data\Managed\UnityEngine.CoreModule.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply Nitrox.Bootloader/Nitrox.Bootloader.csproj]: a379f08

File: NitroxLauncher/NitroxLauncher.csproj [hunk 3]
```diff
@@ -266,19 +266,14 @@
...
-      <Version>2.0.4.0</Version>
-    </PackageReference>
     <PackageReference Include="WindowsAPICodePack-Core">
       <Version>1.1.2</Version>
     </PackageReference>
     <PackageReference Include="WindowsAPICodePack-Shell">
       <Version>1.1.1</Version>
     </PackageReference>
   </ItemGroup>
+  <ItemGroup />
```
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply NitroxLauncher/NitroxLauncher.csproj]: 6f61333

File: NitroxLauncher/BepInEx/BepInEx/config/BepInEx.cfg (hunk 1 above)
Comment (MrPurple6411) [reply NitroxLauncher/BepInEx/BepInEx/config/BepInEx.cfg]: 2 fold.
1. SystemsSpawner is the VERY first monobehavior to run in the game. It is what actually Creates the gameobject with the GameInput monobehavior on it.
2. Currently this is the location QMM will be using and the config is one file that both will use and doesn't really need to be switched out.

File: NitroxLauncher/Patching/QModHelper.cs [hunk 4]
```diff
@@ -6,56 +6,80 @@ namespace NitroxLauncher.Patching
...
+            qmmPatchersOriginPath = patchersOriginPath;
+            qmmPluginBackupPath = Path.Combine(subnauticaBasePath, "QMMBackup", "plugins");
+            qmmPatchersBackupPath = Path.Combine(subnauticaBasePath, "QMMBackup", "patchers");
+            Directory.CreateDirectory(qmmPluginBackupPath);
+            Directory.CreateDirectory(qmmPatchersBackupPath);
+
+
+            Log.Info("Attempting to remove QModManager");
+
+            Exception error = NitroxEntryPatch.RetryWait(() => Directory.Move(qmmPatchersOriginPath, Path.Combine(qmmPatchersBackupPath, "QModManager")), 100, 5) ?? NitroxEntryPatch.RetryWait(() => Directory.Move(qmmPluginOriginPath, Path.Combine(qmmPluginBackupPath, "QModManager")), 100, 5);
```
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply NitroxLauncher/Patching/QModHelper.cs:38]: I mean I can split it into 2 but the jist of it is it will ONLY run the second directory move task if the first one does not produce an error.

File: NitroxLauncher/Patching/NitroxEntryPatch.cs [hunk 5]
```diff
@@ -1,75 +1,112 @@
...
+                if (error != null)
+                {
+                    throw error;
+                }
+            }
+            
+            error = RetryWait(() => Directory.CreateDirectory(nitroxFolder), 100, 5);
+            if (error != null)
+            {
+                Log.Error(error, "Unable to create Nitrox folder.");
```
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply NitroxLauncher/Patching/NitroxEntryPatch.cs:65]: acd1a3a

File: NitroxLauncher/Patching/QModHelper.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply NitroxLauncher/Patching/QModHelper.cs]: 49d043b

File: NitroxLauncher/Patching/NitroxEntryPatch.cs [hunk 6]
```diff
@@ -1,75 +1,112 @@
...
             {
-                TypeDef nitroxMainDefinition = nitroxPatcherAssembly.GetTypes().FirstOrDefault(x => x.Name == NITROX_ENTRY_TYPE_NAME);
-                MethodDef executeMethodDefinition = nitroxMainDefinition.Methods.FirstOrDefault(x => x.Name == NITROX_ENTRY_METHOD_NAME);
-
-                MemberRef executeMethodReference = module.Import(executeMethodDefinition);
+                error = RetryWait(() => QModHelper.RemoveQModManagerFolders(subnauticaCorePath), 100, 5);
+                if (error != null)
+                {
+                    throw error;
+                }
```
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply NitroxLauncher/Patching/NitroxEntryPatch.cs]: yea I guess it somehow could be possible to have qmm files without bepinex but it really shouldn't be possible. however it wont hurt anything to allow it to try anyway as if they are not there then it wont do anything.
116cb5d

File: Nitrox.Bootloader/Main.cs [hunk 7]
```diff
@@ -1,14 +1,20 @@
...
 {
-    public static class Main
+    [BepInPlugin(GUID, MOD_NAME, VERSION)]
+    public class Main: BaseUnityPlugin
     {
-        private static readonly Lazy<string> nitroxLauncherDir = new Lazy<string>(() =>
+        public const string
+            MOD_NAME = "Nitrox",
+            GUID = "com.nitroxmod",
+            VERSION = "1.4.0.0";
```
[the remark this answers is not in the document]
Comment (MrPurple6411) [reply Nitrox.Bootloader/Main.cs:15]: I have tried to find a wait to do this however all I can find are ways to do the opposite. I can make the AssemblyVersion pull from a constant in one of the mods but I cant make a constant pull from anything.

**Comment by Measurity** [comment]
Do we have to bundle Bepinex binaries with the code? If so would be nice to have it in an assets project and not Nitrox Launcher (though the latter could be done in another PR).

**Comment by MrPurple6411** [comment]
I actually already moved it out to the solution directory as requested earlier because 3 projects now use it.

**Comment by Measurity** [comment]
> but yea it does need to go in the build so that the launcher can install it if its not found

If we could download it instead it would be great.
